### DISCUSSION TITLE
XOAUTH2 implementation to Dovecot and Postfix

### DISCUSSION THREAD

**Discussion by alexsalex** [Mailu/Mailu#3681] [Feature requests / Ideas]
I have Mailu server without any web interface for users to check email (like roundcube). And I have RoundCube standslone, on different server. RoundCube natively support OIDC (oAuth2). But this authentication related not only for RoundCube, but for Dovecot and Postfix.

It will really nice to have ability to authenticate with Authentic with RoundCube in to Mailu. Please, implement the XOAUTH2 in to Dovecot and Postfix. Bagging you...

**Comment by alexsalex** [comment]
Enabling XOAUTH2 for both SMTPD and IMAP/POP in Dovecot involves integrating an OAuth2 provider (e.g., Google, Microsoft, or your custom OpenID Connect implementation). Here's how to set it up:

1. Prerequisites
Dovecot version: Ensure you are running Dovecot 2.3+ (XOAUTH2 is supported in this version and above).
OAuth2 Provider: Obtain client credentials (client ID and client secret) from your OAuth2 provider.
SMTP Server: Your SMTP server (e.g., Postfix) must also support XOAUTH2, often configured alongside Dovecot.
Dependencies: Install the liboauth2 plugin for Dovecot.
2. Install Required Packages
Ensure you have the required dependencies installed. On most Linux distributions:

`apt install dovecot-core dovecot-oauth2-plugin dovecot-imapd dovecot-pop3d`

3. Configure OAuth2 in Dovecot
Edit your Dovecot configuration files to enable OAuth2.

Edit **10-auth.conf**

Open the file:
`vi /etc/dovecot/conf.d/10-auth.conf`

Enable XOAUTH2:
```
auth_mechanisms = plain login oauthbearer xoauth2
Ensure authentication is not blocked:
disable_plaintext_auth = no
```
Add OAuth2 Configuration

Create a new configuration file, e.g., **/etc/dovecot/conf.d/10-oauth2.conf**:

Define the OAuth2 client details and token endpoint:
```
passdb {
    driver = oauth2
    mechanisms = xoauth2
    args = /etc/dovecot/dovecot-oauth2.conf.ext
}

userdb {
    driver = static
    args = uid=vmail gid=vmail home=/var/mail/%d/%n
}
```
Create /etc/dovecot/dovecot-oauth2.conf.ext:
```
oauth2_uri = https://accounts.google.com/o/oauth2/token
oauth2_client_id = your-client-id.apps.googleusercontent.com
oauth2_client_secret = your-client-secret
oauth2_username_attr = user
oauth2_token_attr = pass
```
Replace https://accounts.google.com/o/oauth2/token with the token endpoint of your OAuth provider.

Ensure the file permissions are secure:
`chmod 600 /etc/dovecot/dovecot-oauth2.conf.ext`

4. Configure Postfix for XOAUTH2
If you are using Postfix as the SMTP server, configure it to support XOAUTH2.

Edit **main.cf**

Set the authentication mechanism:
```
smtpd_sasl_type = dovecot
smtpd_sasl_path = private/auth
smtpd_sasl_auth_enable = yes
smtpd_recipient_restrictions = permit_sasl_authenticated, permit_mynetworks, reject_unauth_destination
Add the authentication mechanisms:
smtpd_sasl_security_options = noanonymous
smtpd_sasl_mechanism_filter = oauthbearer, xoauth2
```
Edit **master.cf**

Configure the submission service for SASL:
```
submission inet n - n - - smtpd
    -o smtpd_sasl_auth_enable=yes
    -o smtpd_tls_security_level=encrypt
    -o smtpd_tls_wrappermode=no
    -o smtpd_recipient_restrictions=permit_sasl_authenticated,reject
```
5. Test the Setup
Authenticate with XOAUTH2: Use a client like openssl or a compatible mail client to authenticate using XOAUTH2 tokens.
Generate an access token using your OAuth2 provider.
Use the token as the password in the IMAP/SMTP connection.
Enable Debugging: Add the following to **/etc/dovecot/conf.d/10-logging.conf** for troubleshooting:
```
auth_debug = yes
auth_debug_passwords = yes
mail_debug = yes
```
Check the logs:

`tail -f /var/log/mail.log`

6. Additional Notes
Token Refresh: Implement a mechanism to refresh OAuth2 tokens if you're not relying on a client like Thunderbird or Outlook to handle this.
Secure Communication: Ensure both IMAP/SMTP and the OAuth2 communication happen over TLS/SSL.
Firewall: Allow necessary ports (e.g., 993 for IMAP and 587 for SMTP).
Let me know if you encounter specific issues during the setup!

**Reply by nextgens** [reply]
Posting AI-generated bollux will not yield the desired effect.

You may get banned but it will not encourage others to look into what you are asking for.

**Reply by alexsalex** [reply]
I follow this instruction and successful install it on my test server with dovecot without containers.

**Comment by alexsalex** [comment]
https://doc.dovecot.org/2.3/configuration_manual/authentication/oauth2/

**Comment by nickovs** [comment]
This has recently become more pressing because Microsoft have [deprecated (and broken)](https://learn.microsoft.com/en-us/answers/questions/5630452/after-change-to-new-outlook-invalidcredentials-int) BASIC/PLAIN auth in the latest version of Outlook. I have a couple of users who insist on using Outlook and since the latest release they are unable to connect to my MailU instance due to this. I've been after 2FA (#2222) for a while; supporting OAUTH flows would enable this too.

**Reply by nextgens** [reply]
Outlook is broken, news at 11.

From the link you have provided:
_"The reason is that Outlook supports Modern authentication only for Exchange profiles, Outlook.com, and Gmail at this time."_

Whether Mailu (please note the spelling) does or does not support OAuth is irrelevant since there is no way to configure Outlook to use it.
